### ISSUE TITLE
task1_ai_ml_model : Initial research on Conversational algorithms for PTSD risk/assesment - seq2seq

### ISSUE THREAD

**Issue by nlauchande** [omdena/ptsd#5]
Goals ( WIP)  :

1. Learn about conversational systems - We need some resources for it. My suggestion at the moment is to look at the initial videos of http://web.stanford.edu/class/cs224u/ . 

2. Work on a prototype on one of the tools investigated by the team implement a dummy conversational therapist based on the pdf provided. (edited) 

3. As part of this subtask we will also need to work on https://github.com/omdena/ptsd/issues/3 or collaborate with the datasets team to help us sorting this out.

Listing of resources being studied (not exhaustive) : 
1.  https://arxiv.org/pdf/1809.00640.pdf#page11 - suggested by @elateifsara 
     Summary by Rishika 
     https://docs.google.com/document/d/1jUlXGnZZlfWpdQUWWF_NIO8GEdDurcCAnF9e6wDkyGA/edit

**Comment by sylvia31096** [comment]
> I do not know if this can help out. But this site has examples of transcripts between patients and therapists.
> https://www.apa.org/ptsd-guideline/resources/case-examples

Using the case example transcripts, I was thinking we could work with psychologists to figure out what are the key issues they pick up from clients in a session and maybe that could form our entities in RASA.

**Comment by BottleSpink** [comment]
1. So, the first subtask on the conversational component is to research about different conversational This will be for overall conversational approach, including diagnosis, treatment or be it any intent. Can we see this as dialog management? To start with, this week, we would like to work on the prototype as in, a basic conversations builder with users.

2. The second task is specific to NLU algorithms in diagnosis.

My understanding: NLP (With respect to chatbot)= NLU(understanding)+NLG(generation)

NLU is mostly to do with extracting the intent and  entity (What has been said/asked by the user) and NLG is something like dialogue generation (what the bot has to reply) so I see subtask 1 as NLG and subtask 2 as NLU (diagnosis)

I am planning to look at seq2seq architecture. There are chatbots built end-to-end based on this model. So for this week, we (in subtask1) can look at this architecture and work towards building a very basic conversational algorithm

**Comment by sylvia31096** [comment]
My POV for the treatment was more of like the therapy session number or level would form the intent e.g session 1 could be to describe the event, then the entity could be the emotions ( or whatever therapists would like to extract) and then the bot would give a response in regards to that. Would that work?

**Comment by albertlai431** [comment]
From a basic viewpoint, the seq2seq model consists of 3 parts:

Encoder: bidirectional LSTM runs through the input sequence one word at a time, outputting a hidden state and output vector. The hidden state vector is passed onto the next time step as input, and the output vector is recorded to the encoder state/"context". 

Attention mechanism: basically tells the decoder which part of the sentence to focus on when generating a sequence.

Decoder: Takes context vectors, hidden states, and outputs from the attention mechanism to generate the sequence to be output to the user.

**Comment by nlauchande** [comment]
@sylvia31096 : 

This issue, in particular, is just to research techniques and potentially nlu + nlg algorithms with a focus on treatment/therapy .

**We want to synthesize a virtual/chatbot therapist for PSTD patients based on dialogs of therapy.**

**Comment by kulsoom-abdullah** [comment]
I read the paper - Deep learning for language understanding of mental health concepts derived from Cognitive Behavioural Therapy

I was first interested in the data they used which was 500K posts in an online platform. These posts were a negative take on a problem, and peers respond with a positive rethink. The responses were not collected for the dataset.

Labeling of a subset of the data was performed by two psychological therapists.  They annotated posts by adding or removing the labels in the categories 1) thinking errors, 2) emotions and 3) situations.  

Here is one of the results:
Problem: I really miss my ex! She just stopped talking to me all of the sudden, I don't know what happened
Negative Take: I wont get to talk to her again
Thinking Errors                                    Emotions   Situation
     True LABELS
- Jumping to negative conclusions    - Hurt        - Relationship
- Low Frustration Tolerance                - Grief       - Relationship

A question I have is this similar to the goal of the project - that is, to take words of someone and detect or classify something related to PTSD and guide them where to get help? I want to make sure I understand.  Thanks

**Comment by BottleSpink** [comment]
I have built a basic conversational bot using the tutorial given in the pytorch tutorial site. It uses seq2seq. To start with, it uses movie corpus and it is mostly retrieval based model. But it is a good learning till now. Next steps: 
1. To look at how I can bring in our transcript dataset
2. To improve the model by building a generative model

**Comment by sylvia31096** [comment]
@BottleSpink could you push the code?

**Comment by BottleSpink** [comment]
> @BottleSpink could you push the code?

Sure, will do! I will do some cleanup and push the notebook file

**Comment by nlauchande** [comment]
@BottleSpink: Please looking to look at your code whenever you can.

**Comment by albertlai431** [comment]
I believe that after training the model, we can put the .tar file into the models folder and switch between the different models we have depending on the diagnosis or the therapy!

**Comment by SubmitCode** [comment]
Here are my thoughts after reviewing some papers and testing rasa:

- **Small Data Problem** it seems like that we don't have much data and even if we get our hands on it might be different from the actual conversations therefore I thought we should build our conversational bot with an additional Admin UI which can be used by experts to label new conversations and maybe add intends and stories through a website so it becomes better and better over time but might not be as good at the beginning. I found RASA X provides something like this.
- **Diagnosis of PTSD** should be our main focus but it should happen through conversation not just multiple choice questionaires
- **Treatment of PTSD** should be secondary as it much more difficult and it might have many nuances.
- **RASA** is my favourite for several reasons but mainly because it's open source and free so you only need a host and you are independent. (for RASA X I am not sure if it's an addon which is just not open source or if it's the complet RASA with added functionality)

**Comment by nlauchande** [comment]
Spent a good deal of time studying SEQ2SEQ models. As suggested in many of the channels: https://pytorch.org/tutorials/beginner/chatbot_tutorial.html .  My notes so far :

1. Replicated the tutorial in Google Collab 

2. Seq2Seq sounds like a good option if you want to generate conversational bots out of transcripts of Therapy sessions. In the case, we had appropriate data.

3. Busy finalizing running the same tutorial with the PTSD tutorials available. I don't expect the results to be good but will allow the production of a standard format for the data.

Will update a github branch with the initial code ready to be tested.

**Comment by sylvia31096** [comment]
Was there any discussion yesterday on the conversational part of the chatbot? Is it still a route we are considering?
 I have also been trying to replicate a tutorial on seq2seq in Colab.

**Comment by nlauchande** [comment]
@sylvia31096  :
We are definitely keen to carry on with the conversational research but focusing a bit more on Diagnostics. Wonder if we should change the name of this issue and the goals to reflect that intent.

**Comment by nlauchande** [comment]
Closing this issue for now. If anyone has strong feeling about it please let's reopen it.
